From patchwork Mon Jan 28 17:43:22 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Anton Kuchin <antonkuchin@yandex-team.ru>
X-Patchwork-Id: 10784175
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id C2A4D6C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 17:51:42 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AFA352BEB9
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 17:51:42 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id A2E7A2BE91; Mon, 28 Jan 2019 17:51:42 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham
	version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 209562BE82
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 17:51:42 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727542AbfA1Rvk (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 28 Jan 2019 12:51:40 -0500
Received: from forwardcorp1o.cmail.yandex.net ([37.9.109.47]:53483 "EHLO
        forwardcorp1o.cmail.yandex.net" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with ESMTP id S1727502AbfA1Rvj (ORCPT
        <rfc822;kvm@vger.kernel.org>); Mon, 28 Jan 2019 12:51:39 -0500
X-Greylist: delayed 486 seconds by postgrey-1.27 at vger.kernel.org;
 Mon, 28 Jan 2019 12:51:38 EST
Received: from mxbackcorp1o.mail.yandex.net (mxbackcorp1o.mail.yandex.net
 [IPv6:2a02:6b8:0:1a2d::301])
        by forwardcorp1o.cmail.yandex.net (Yandex) with ESMTP id CD98021804;
        Mon, 28 Jan 2019 20:43:30 +0300 (MSK)
Received: from smtpcorp1p.mail.yandex.net (smtpcorp1p.mail.yandex.net
 [2a02:6b8:0:1472:2741:0:8b6:10])
        by mxbackcorp1o.mail.yandex.net (nwsmtp/Yandex) with ESMTP id
 WpIe7eNbvd-hUAm5q0P;
        Mon, 28 Jan 2019 20:43:30 +0300
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=yandex-team.ru;
 s=default;
        t=1548697410; bh=CFh/ujDWHZz9OPMECaaUL/o5Z/rulgzr5IhbDE9w0ow=;
        h=From:To:Cc:Subject:Date:Message-Id;
        b=leemMJ2AqQ/4HO8+zQeBhc/EvSXVOL5jSPGpbYBnMfjBhOosmQeKZGyDWNwTltGu3
         oSWAb/Y9vkmZmY2y318KFe9QL4kvLSsb+6zass4bQpeUbyWVN+DmWpuaTw5pfxEIle
         LsmN5G8MxQLvWhVzWhzbeQ2Qubnv67PySg85UZPY=
Authentication-Results: mxbackcorp1o.mail.yandex.net;
 dkim=pass header.i=@yandex-team.ru
Received: from dynamic-red.dhcp.yndx.net (dynamic-red.dhcp.yndx.net
 [2a02:6b8:0:408:250:b6ff:fe97:2682])
        by smtpcorp1p.mail.yandex.net (nwsmtp/Yandex) with ESMTPSA id
 VJGuASzpeM-hU5O80gK;
        Mon, 28 Jan 2019 20:43:30 +0300
        (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
        (Client certificate not present)
From: Anton Kuchin <antonkuchin@yandex-team.ru>
To: kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Borislav Petkov <bp@alien8.de>, x86@kernel.org,
 "H. Peter Anvin" <hpa@zytor.com>, linux-kernel@vger.kernel.org,
 Thomas Gleixner <tglx@linutronix.de>, Ingo Molnar <mingo@redhat.com>,
 Evgeny Yakovlev <wrfsh@yandex-team.ru>,
 Anton Kuchin <antonkuchin@yandex-team.ru>
Subject: [PATCH] KVM: Ignore LBR MSRs with no effect
Date: Mon, 28 Jan 2019 20:43:22 +0300
Message-Id: <20190128174322.8529-1-antonkuchin@yandex-team.ru>
X-Mailer: git-send-email 2.19.1
MIME-Version: 1.0
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Win10 attempts to save these registers during KiSaveDebugRegisterState
if LBR or BTF bits are set in MSR_IA32_DEBUGCTLMSR. It uses DR7 GE and LE
flags for per-thread switching of these these features so zero value that
is returned for MSR_IA32_DEBUGCTLMSR has no effect.

These registers are used for debugging and shouldn't cause #GP and
guest crash so just return zeroes just like we do for common x86 LBR
MSRs (DEBUGCTLMSR, LAST[BRANCH|INT][TO|FROM]IP).

Signed-off-by: Anton Kuchin <antonkuchin@yandex-team.ru>
---
 arch/x86/kvm/vmx/vmx.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index f6915f10e584..8bc56cf027ed 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -1769,6 +1769,13 @@ static int vmx_get_msr(struct kvm_vcpu *vcpu, struct msr_data *msr_info)
 		else
 			msr_info->data = vmx->pt_desc.guest.addr_a[index / 2];
 		break;
+	case MSR_LBR_TOS:
+	case MSR_LBR_NHM_FROM ... MSR_LBR_NHM_FROM + 31:
+	case MSR_LBR_NHM_TO ... MSR_LBR_NHM_TO + 31:
+	case MSR_LBR_CORE_FROM ... MSR_LBR_CORE_FROM + 7:
+	case MSR_LBR_CORE_TO ... MSR_LBR_CORE_TO + 7:
+		msr_info->data = 0;
+		break;
 	case MSR_TSC_AUX:
 		if (!msr_info->host_initiated &&
 		    !guest_cpuid_has(vcpu, X86_FEATURE_RDTSCP))
